buffer/src/main/java/io/netty/buffer/AbstractReferenceCountedByteBuf.java

```diff
@@ -83,11 +93,18 @@ public ByteBuf retain(int increment) {
     }
 
     private ByteBuf retain0(final int increment) {
-        int oldRef = refCntUpdater.getAndAdd(this, increment);
-        if (oldRef <= 0 || oldRef + increment < oldRef) {
-            // Ensure we don't resurrect (which means the refCnt was 0) and also that we encountered an overflow.
+        // all changes to the raw count are 2x the "real" change
+        int adjustIncrement = increment << 1; // overflow OK here
+        int oldRef = refCntUpdater.getAndAdd(this, adjustIncrement);
+        if ((oldRef & 1) != 0) {
+            throw new IllegalReferenceCountException(0, increment);
+        }
+        // don't pass 0!
+        if ((oldRef <= 0 && oldRef + adjustIncrement >= 0)
+                || (oldRef >= 0 && oldRef + adjustIncrement < oldRef)) {
+            // overflow case
             refCntUpdater.getAndAdd(this, -increment);
```

Comment:
does this need to be adjustIncrement?

Reply:
good spot, thanks @carl-mastrangelo, will fix